flink-formats/flink-json/src/main/java/org/apache/flink/formats/json/JsonFileSystemFormatFactory.java

```diff
@@ -0,0 +1,293 @@
+		RowType nonPartRowType = RowType.of(
+			Arrays.asList(nonPartFieldTypes).stream()
+				.map(DataType::getLogicalType)
+				.toArray(LogicalType[]::new),
+			nonPartFieldNames);
+
+		// deal project fields
+		int[] selectFields = context.getProjectFields();
+		List<String> partitionKeys = context.getPartitionKeys();
+		List<String> selectFieldNames = Arrays.stream(selectFields)
+			.mapToObj(fieldNames::get)
+			.collect(Collectors.toList());
+		List<String> jsonSelectFieldNames = selectFieldNames.stream()
+			.filter(name -> !partitionKeys.contains(name)).collect(Collectors.toList());
+		int[] jsonFieldMapping =  jsonSelectFieldNames.stream().mapToInt(selectFieldNames::indexOf).toArray();
+
+		JsonRowDataDeserializationSchema  deserializationSchema = new JsonRowDataDeserializationSchema(
+			nonPartRowType,
+			new GenericTypeInfo(GenericRowData.class),
+			failOnMissingField,
+			ignoreParseErrors);
+
+		return new JsonFileSystemFormat(
+			context.getPaths(),
+			fieldTypes,
+			fieldNames,
+			selectFields,
+			partitionKeys,
+			context.getDefaultPartName(),
+			context.getPushedDownLimit(),
+			jsonSelectFieldNames,
+			jsonFieldMapping,
+			deserializationSchema);
+	}
+
+	@Override
+	public Optional<Encoder<RowData>> createEncoder(WriterContext context) {
+		TableSchema nonPartKeySchema = new TableSchema.Builder()
+			.fields(context.getFieldNamesWithoutPartKeys(), context.getFieldTypesWithoutPartKeys())
+			.build();
+		LogicalType[] logicalTypes = Arrays.asList(nonPartKeySchema.getFieldDataTypes())
+			.stream().map(DataType::getLogicalType).toArray(LogicalType[]::new);
+		RowType nonPartRowType = RowType.of(logicalTypes, nonPartKeySchema.getFieldNames());
+		return Optional.of(new JsonRowDataEncoder(new JsonRowDataSerializationSchema(nonPartRowType)));
+	}
+
+	@Override
+	public Optional<BulkWriter.Factory<RowData>> createBulkWriterFactory(WriterContext context) {
+		return Optional.empty();
+	}
+
+	@Override
+	public boolean supportsSchemaDerivation() {
+		return true;
+	}
+
+	private static DescriptorProperties getValidatedProperties(Map<String, String> propertiesMap) {
+		final DescriptorProperties properties = new DescriptorProperties(true);
+		properties.putProperties(propertiesMap);
+		properties.validateBoolean(FORMAT_FAIL_ON_MISSING_FIELD, true);
+		properties.validateBoolean(FORMAT_IGNORE_PARSE_ERRORS, true);
+		return properties;
+	}
+
+	/**
+	 * A {@link JsonFileSystemFormat} is responsible to read {@link RowData} records
+	 * from json format files.
+	 */
+	public static class JsonFileSystemFormat extends DelimitedInputFormat<RowData> {
+		/**
+		 * Code of \r, used to remove \r from a line when the line ends with \r\n.
+		 */
+		private static final byte CARRIAGE_RETURN = (byte) '\r';
+
+		/**
+		 * Code of \n, used to identify if \n is used as delimiter.
+		 */
+		private static final byte NEW_LINE = (byte) '\n';
+
+		/**
+		 * The name of the charset to use for decoding.
+		 */
+		private String charsetName = "UTF-8";
```

Comment:
What is this for?

Reply:
I wanted to construct a transient Charset, but now it don't need any more after use StandardCharsets.UTF_8